File: detection-engine/internal/rules/event.go
// Package rules содержит типы событий и структуры правил детектирования.
package rules

import "time"

// Event — нормализованное событие из Kafka (соответствует NormalizedEvent из Rust).
type Event struct {
	Timestamp  time.Time         `json:"@timestamp"`
	EventID    string            `json:"event_id"`
	SourceType string            `json:"source_type"`
	EventType  string            `json:"event_type"`
	Severity   string            `json:"severity"`
	Message    string            `json:"message"`
	Host       string            `json:"host"`
	SourceIP   *string           `json:"source_ip"`
	UserID     *string           `json:"user_id"`
	Action     *string           `json:"action"`
	StatusCode *int              `json:"status_code"`
	URLPath    *string           `json:"url_path"`
	HTTPMethod *string           `json:"http_method"`
	DurationMs *float64          `json:"duration_ms"`
	Metadata   map[string]any    `json:"metadata"`
}

// StrVal возвращает строковое значение указателя или пустую строку.
func StrVal(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}

// IntVal возвращает int значение указателя или 0.
func IntVal(i *int) int {
	if i == nil {
		return 0
	}
	return *i
}

File: detection-engine-rs/src/event.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Event {
    #[serde(rename = "@timestamp")]
    pub timestamp: DateTime<Utc>,
    pub event_id: String,
    pub source_type: String,
    pub event_type: String,
    pub severity: String,
    pub message: String,
    pub host: String,
    #[serde(default)]
    pub source_ip: Option<String>,
    #[serde(default)]
    pub user_id: Option<String>,
    #[serde(default)]
    pub action: Option<String>,
    #[serde(default)]
    pub status_code: Option<i32>,
    #[serde(default)]
    pub url_path: Option<String>,
    #[serde(default)]
    pub http_method: Option<String>,
    #[serde(default)]
    pub duration_ms: Option<f64>,
    #[serde(default)]
    pub metadata: HashMap<String, serde_json::Value>,
}

impl Event {
    pub fn str_val(opt: &Option<String>) -> &str {
        opt.as_deref().unwrap_or("")
    }

    pub fn int_val(opt: &Option<i32>) -> i32 {
        opt.unwrap_or(0)
    }
}
